trl/agents/utils.py

```diff
@@ -0,0 +1,270 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright 2025 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# from __future__ import annotations
+
+import asyncio
+from inspect import getsource
+from pathlib import Path
+from typing import Callable
+
+from ..import_utils import is_e2b_available, is_langchain_experimental_available
+
+
+if is_e2b_available():
+    from e2b_code_interpreter import AsyncSandbox  # used for E2B async code execution
+
+if is_langchain_experimental_available():
+    from langchain_experimental.utilities import PythonREPL  # used for local code execution
+
+default_system_prompt = "You can answer questions and solve problems. If running code helps, write it inside <code> </code>, and you will see the result. Example: To calculate 2 + 2, write <code> print(2 + 2) </code>."
+
+default_environment_prompt = "This is a user-provided script containing tools that you can use to help complete tasks. It will be added to the sandbox environment so you can call its functions. If you are unsure how to use the available tools, you can use the `help()` function to inspect them."
+
+
+def read_script(user_script_path: str) -> str:
+    """
+    Reads and returns the content of the provided script.
+
+    Args:
+        user_script_path (`str`):
+            Path to the user-provided script.
+
+    Returns:
+        `str`:
+            The content of the script.
+    """
+    return Path(user_script_path).read_text()
+
+
+class Localexecuter:
+    def execute(self, codes: list[str]) -> list[str]:
+        """
+        Executes multiple code snippets using PythonREPL sequentially.
+
+        Args:
+            codes (list[str]): list of code snippets to execute.
+
+        Returns:
+            list[str]: list of execution results in same order as input snippets.
+        """
+        results = []
+        repl = PythonREPL()
+        for code in codes:
+            try:
+                result = repl.run(code)
+                results.append(str(result))
+            except Exception as e:
+                results.append(f"Error executing code: {str(e)}")
+        return results
+
+
+class E2BExecuter:
```

Comment:
turbo nit:
```suggestion
class E2BExecutor:
```

Similar question here on whether we can reuse the `smolagents` one: https://github.com/huggingface/smolagents/blob/d8db112be8f208db851c0836d56de0c9b157cd6d/src/smolagents/remote_executors.py#L96

Reply:
Not really, the biggest hurdle in integrating smolagents into TRL, as I explained before, is running things in batches and in parallel.
This E2BCodeExecuter uses asyncio to run batch of code but smolagents doesn't.

Reply:
OK then let's stick with your implementation for now and later if `smolagents` supports batching / multiprocessing, we can revisit